vnext/JSI.Desktop.UnitTests/JsiRuntimeUnitTests.cpp

```diff
@@ -0,0 +1,1072 @@
+    Value get(Runtime& rt, const PropNameID& sym) override {
+      return String::createFromUtf8(rt, sym.utf8(rt) + sym.utf8(rt));
+    }
+
+    void set(Runtime&, const PropNameID&, const Value&) override {}
+  };
+
+  Object tho =
+    Object::createFromHostObject(rt, std::make_shared<TwiceHostObject>());
+  EXPECT_TRUE(function("function (obj) { return obj.abc == 'abcabc'; }")
+    .call(rt, tho)
+    .getBool());
+  EXPECT_TRUE(function("function (obj) { return obj['def'] == 'defdef'; }")
+    .call(rt, tho)
+    .getBool());
+  EXPECT_TRUE(function("function (obj) { return obj[12] === '1212'; }")
+    .call(rt, tho)
+    .getBool());
+  EXPECT_TRUE(tho.isHostObject(rt));
+  EXPECT_TRUE(
+    std::dynamic_pointer_cast<ConstantHostObject>(tho.getHostObject(rt)) ==
+    nullptr);
+  EXPECT_TRUE(tho.getHostObject<TwiceHostObject>(rt).get() != nullptr);
+
+  class PropNameIDHostObject : public HostObject {
+    Value get(Runtime& rt, const PropNameID& sym) override {
+      if (PropNameID::compare(rt, sym, PropNameID::forAscii(rt, "undef"))) {
+        return Value::undefined();
+      }
+      else {
+        return PropNameID::compare(
+          rt, sym, PropNameID::forAscii(rt, "somesymbol"));
+      }
+    }
+
+    void set(Runtime&, const PropNameID&, const Value&) override {}
+  };
+
+  Object sho = Object::createFromHostObject(
+    rt, std::make_shared<PropNameIDHostObject>());
+  EXPECT_TRUE(sho.isHostObject(rt));
+  EXPECT_TRUE(function("function (obj) { return obj.undef; }")
+    .call(rt, sho)
+    .isUndefined());
+  EXPECT_TRUE(function("function (obj) { return obj.somesymbol; }")
+    .call(rt, sho)
+    .getBool());
+  EXPECT_FALSE(function("function (obj) { return obj.notsomuch; }")
+    .call(rt, sho)
+    .getBool());
+
+  class BagHostObject : public HostObject {
+  public:
+    const std::string& getThing() {
+      return bag_["thing"];
+    }
+
+  private:
+    Value get(Runtime& rt, const PropNameID& sym) override {
+      if (sym.utf8(rt) == "thing") {
+        return String::createFromUtf8(rt, bag_[sym.utf8(rt)]);
+      }
+      return Value::undefined();
+    }
+
+    void set(Runtime& rt, const PropNameID& sym, const Value& val) override {
+      std::string key(sym.utf8(rt));
+      if (key == "thing") {
+        bag_[key] = val.toString(rt).utf8(rt);
+      }
+    }
+
+    std::unordered_map<std::string, std::string> bag_;
+  };
+
+  std::shared_ptr<BagHostObject> shbho = std::make_shared<BagHostObject>();
+  Object bho = Object::createFromHostObject(rt, shbho);
+  EXPECT_TRUE(bho.isHostObject(rt));
+  EXPECT_TRUE(function("function (obj) { return obj.undef; }")
+    .call(rt, bho)
+    .isUndefined());
+  EXPECT_EQ(
+    function("function (obj) { obj.thing = 'hello'; return obj.thing; }")
+    .call(rt, bho)
+    .toString(rt)
+    .utf8(rt),
+    "hello");
+  EXPECT_EQ(shbho->getThing(), "hello");
+
+  // TODO (T28293178) Remove this once exceptions are supported in all builds.
+#ifndef JSI_NO_EXCEPTION_TESTS
```

Comment:
>JSI_NO_EXCEPTION_TESTS [](start = 8, length = 22)

Should this be moved into a separate test file?

Reply:
We chatted offline about this.